src/DocumentFormat.OpenXml/Framework/Schema/ParticlePath.cs

```diff
@@ -0,0 +1,135 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using DocumentFormat.OpenXml.Validation.Schema;
+using System;
+
+namespace DocumentFormat.OpenXml.Framework.Schema
+{
+    internal class ParticlePath : IComparable<ParticlePath>, IEquatable<ParticlePath>
+    {
+        private readonly ParticlePathItem[] _values;
+
+        public static ParticlePath Create(params ParticlePathItem[] values)
+            => new ParticlePath(values);
+
+        private ParticlePath(ParticlePathItem[] values)
+        {
+            _values = values;
+        }
+
+        public ParticleType Type => _values[_values.Length - 1].Type;
+
+        public int CompareTo(ParticlePath other)
+            => CompareTo(other, true);
+
+#if DEBUG
+        public string Path => string.Join(", ", (object[])_values);
+#endif
+
+        private int CompareTo(ParticlePath other, bool isCompare)
+        {
+            if (other is null)
+            {
+                return -1;
+            }
+
+            var en = new Enumerator(this);
+            var on = new Enumerator(other);
+
+            while (en.MoveNext() && on.MoveNext())
+            {
+                var (currentParticleType, currentIndex) = en.Current;
+                var (otherParticleType, otherIndex) = on.Current;
+
+                if (currentParticleType == ParticleType.Element && otherParticleType == ParticleType.Element)
+                {
+                    return currentIndex.CompareTo(otherIndex);
+                }
+                else if (currentParticleType == ParticleType.Sequence && otherParticleType == ParticleType.Sequence)
+                {
+                    var compared = currentIndex.CompareTo(otherIndex);
+
+                    if (compared != 0)
+                    {
+                        return compared;
+                    }
+                }
+                else if (currentParticleType == ParticleType.Choice && otherParticleType == ParticleType.Choice)
+                {
+                    if (currentIndex != otherIndex)
+                    {
+                        return 0;
+                    }
+                }
+                else if (currentParticleType == ParticleType.All && otherParticleType == ParticleType.All)
+                {
+                    var compared = currentIndex.CompareTo(otherIndex);
+
+                    if (compared != 0)
+                    {
+                        return isCompare ? 0 : -1;
+                    }
+                }
+                else
+                {
+                    return 0;
+                }
+            }
+
+            return 0;
+        }
+
+        public bool IsSibling(ParticlePath other)
+          => Equals(other, _values.Length - 1);
+
+        public override bool Equals(object obj)
+            => Equals(obj as ParticlePath);
+
+        public bool Equals(ParticlePath other)
+            => CompareTo(other, false) == 0;
```

Comment:
If the other path has different particle types, `CompareTo` will return 0 (which sort of makes sense since I think there's no reasonable ordering of different types), but that makes `CompareTo` different from `Equals`.

It seems like comparing lengths and calling `Equals(other, length)` if they have the same length would be better. That also has the benefit of simplifying `CompareTo` by removing the need for its second parameter.

Reply:
Equality of particle paths is defined as if they can be replaced. I've added a comment to that effect.